### PR TITLE
feat(relay): Upgrade to version 4

### LINKED ISSUE #95: Upstream RelayCompilerPublic change, rename to index.js
Flagging an issue for future reference, this was merged recently into `relay`:
https://github.com/facebook/relay/pull/2677

This will break importing here: https://github.com/relay-tools/relay-compiler-language-typescript/blob/master/src/TypeScriptGenerator.ts#L6

Happy to open PR when a new release is cut with this change.

### LINKED ISSUE #111: remove graphql-compiler references
graphql-compiler was removed on v2 https://github.com/facebook/relay/releases/tag/v2.0.0

### PR DESCRIPTION
**Pull request by sibelius**
- remove graphql-compiler references and types, fix #111
- fix RelayCompilerPublic change, fix #95
- update transforms based on https://github.com/relay-tools/relay-compiler-language-typescript/pull/109
- update relay version based on https://github.com/relay-tools/relay-compiler-language-typescript/pull/106
- fix facebook/relay#2717
- fix facebook/relay#2710

### CODE REVIEW

File: test/__snapshots__/TypeScriptGenerator-test.ts.snap [hunk 1]
```diff
@@ -433,9 +440,15 @@ export type NameRendererQueryVariables = {
 };
 export type NameRendererQueryResponse = {
     readonly me: {
-        readonly nameRenderer: {
-            readonly " $fragmentRefs": PlainUserNameRenderer_name$ref & MarkdownUserNameRenderer_name$ref;
-        } | null;
```
Comment (alloy) [inline test/__snapshots__/TypeScriptGenerator-test.ts.snap]: Why did these fragment references go away?

File: test/__snapshots__/TypeScriptGenerator-test.ts.snap [hunk 2]
```diff
@@ -1397,9 +1497,15 @@ type PlainUserNameRenderer_name$ref = any;
 export type NameRendererFragment$ref = any;
 export type NameRendererFragment = {
     readonly id: string;
-    readonly nameRenderer: {
-        readonly " $fragmentRefs": PlainUserNameRenderer_name$ref & MarkdownUserNameRenderer_name$ref;
-    } | null;
```
Comment (alloy) [inline test/__snapshots__/TypeScriptGenerator-test.ts.snap]: Ditto

File: test/__snapshots__/TypeScriptGenerator-test.ts.snap [hunk 3]
```diff
@@ -1458,9 +1564,15 @@ export type NameRendererQueryVariables = {
 };
 export type NameRendererQueryResponse = {
     readonly me: {
-        readonly nameRenderer: {
-            readonly " $fragmentRefs": PlainUserNameRenderer_name$ref & MarkdownUserNameRenderer_name$ref;
-        } | null;
```
Comment (alloy) [inline test/__snapshots__/TypeScriptGenerator-test.ts.snap]: Ditto

**Comment by alloy** [comment]
Great stuff, thanks @sibelius 🙌 

> fix facebook/relay#2717
> fix facebook/relay#2710

By these you mean that the changes implicitly fix these issues, not that you made any additional changes in the patch to address them specifically, right?

File: src/TypeScriptGenerator.ts [hunk 4]
```diff
@@ -593,8 +587,12 @@ function getRefTypeName(name: string): string {
   return `${name}$ref`;
 }
 
+// Should match FLOW_TRANSFORMS array
+// https://github.com/facebook/relay/blob/master/packages/relay-compiler/language/javascript/RelayFlowGenerator.js#L660
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: Best to not link to a branch, as that can go out of date:

```suggestion
// https://github.com/facebook/relay/blob/v4.0.0/packages/relay-compiler/language/javascript/RelayFlowGenerator.js#L621-L627
```

**Comment by alloy** [comment]
@zephraph I kinda would want to bump the version to v4. Is there any way to do that with auto-release or should I do that manually?

File: test/__snapshots__/TypeScriptGenerator-test.ts.snap (hunk 1 above)
Comment (sibelius) [reply test/__snapshots__/TypeScriptGenerator-test.ts.snap]: missing more stuff https://github.com/facebook/relay/blob/27377d38189ed12f6658c4307156ec51c32cea8c/packages/relay-compiler/language/javascript/__tests__/__snapshots__/RelayFlowGenerator-test.js.snap#L372

**Comment by alloy** [comment]
Filed an issue for the missing type information when `@match` is used https://github.com/relay-tools/relay-compiler-language-typescript/issues/113, because I think we can still release this patch as-is and iterate on that issue.

**Comment by alloy** [comment]
Released as v4.1.0 (had to do some release faffing that led to skipping v4.0.0) and also just updated in our app and only seeing expected artefact changes, thus far. Great job, thanks again! 👏

File: test/__snapshots__/TypeScriptGenerator-test.ts.snap (hunk 1 above)
Comment (alloy) [reply test/__snapshots__/TypeScriptGenerator-test.ts.snap]: https://github.com/relay-tools/relay-compiler-language-typescript/issues/113

**Comment by jstejada** [comment]
thanks for upgrading this!

**Pull request merged**